src/project/project.ts

```diff
@@ -107,9 +107,13 @@ export class Project {
 
     readonly options: ProjectOptions;
 
-    constructor(config: ProjectConfig | SourceLike[]) {
-        if (Array.isArray(config)) {
-            config = { sources: config };
+    constructor(configOrSources: ProjectConfig | ReadonlyArray<SourceLike>) {
+        let config: ProjectConfig;
+        if (Array.isArray(configOrSources)) {
+            config = { sources: configOrSources };
+        } else {
+            // assertion needed because of https://github.com/microsoft/TypeScript/issues/17002
+            config = configOrSources as ProjectConfig;
         }
```

Comment:
Uff the linked issue is from 2017 and contains ~50 replies 😅 . Who would have thought that such an issue is that complicated 😅 

You can write a type guard for correct narrowing.
```typescript
export function isReadonlyArray<T>(obj: unknown | ReadonlyArray<T>): obj is ReadonlyArray<T> {
   return Array.isArray(obj);
}
```

Reply:
yeah it's really bad. Most of the cases you don't even realize and just get an `any`-typed value that propagates. Should indeed write this helper and use it everywhere